File: api/app/service/file.py
import os
import logging
from fastapi import UploadFile
from corepy.datetime import format_date_compact
from corepy.codec import file_md5
from uuid import uuid4
from pathlib import Path
import shutil

from app.config import open_session, UPLOAD_TEMP_DIR, STATIC_FILE_DIR, FILE_URL_PREFX
from app.repo.file import create_file, get_file

logger = logging.getLogger(__name__)


def upload_file(file: UploadFile) -> str: 
    now_date_str = format_date_compact()
    file_dir = f"{UPLOAD_TEMP_DIR}/{now_date_str}"
    if not os.path.exists(file_dir):
        os.makedirs(file_dir, exist_ok=True)
    filename = file.filename
    if not filename:
        raise Exception("unsupported file")
    suffix = Path(filename).suffix
    size = file.size or 0

    file_temp_path = f"{file_dir}/{str(uuid4()).replace('-', '')}{suffix}"
    print(f"saving upload file {filename} to {file_temp_path}")
    with open(file_temp_path, "wb") as buffer:
        shutil.copyfileobj(file.file, buffer)

    id = file_md5(file_temp_path)

    with open_session() as session:
        pass
    file_entity = get_file(session, id)
    if file_entity:
        logger.warning(f"file {id} already exists, returning existing url")
        return file_entity.url

    static_file_dir = f"{STATIC_FILE_DIR}/{now_date_str}"
    if not os.path.exists(static_file_dir):
        os.makedirs(static_file_dir, exist_ok=True)

    static_file_name = f"{id}{suffix}"
    static_file_path = f"{static_file_dir}/{static_file_name}"
    logger.info(f"moving upload file {file_temp_path} to {static_file_path}")
    shutil.move(file_temp_path, static_file_path)

    url = f"{FILE_URL_PREFX}/{now_date_str}/{static_file_name}"
    file_entity = create_file(session, id, filename, size, url)
    return url

File: api/src/service/file.rs
use std::net::SocketAddr;
use std::path::{Path, PathBuf};

use axum::extract::Multipart;
use chrono::Local;
use tokio::fs;
use tokio::io::AsyncWriteExt;
use tracing::{info, warn};

use crate::config::{ApiState, STATIC_FILE_DIR, UPLOAD_TEMP_DIR};
use crate::dao::file as dao;
use crate::entity::{File, HostingFile};
use crate::model::HostingUploadParam;
use crate::util::{AppError, AppResult, file_md5, format_date_compact, uuid_no_hyphen};

pub struct FileService;

impl FileService {
    pub async fn upload_file(
        state: &ApiState, mut multipart: Multipart, addr: SocketAddr,
    ) -> AppResult<String> {
        let mut filename: Option<String> = None;
        let mut data: Option<Vec<u8>> = None;

        while let Some(field) = multipart
            .next_field()
            .await
            .map_err(|e| AppError::biz(format!("invalid multipart: {}", e)))?
        {
            if field.name() == Some("file") {
                filename = field.file_name().map(|s| s.to_string());
                data = Some(
                    field
                        .bytes()
                        .await
                        .map_err(|e| AppError::biz(format!("read upload data failed: {}", e)))?
                        .to_vec(),
                );
                break;
            }
        }

        let filename = filename.ok_or_else(|| AppError::biz("unsupported file"))?;
        let data = data.ok_or_else(|| AppError::biz("unsupported file"))?;
        let size = data.len() as i64;

        println!("[{}] POST /api/v1/upload file={} size={}", addr, filename, size);

        let now_date_str = format_date_compact();
        let file_dir = format!("{}/{}", UPLOAD_TEMP_DIR, now_date_str);
        fs::create_dir_all(&file_dir).await?;

        let suffix = Path::new(&filename)
            .extension()
            .map(|s| format!(".{}", s.to_string_lossy()))
            .unwrap_or_default();

        let temp_name = format!("{}{}", uuid_no_hyphen(), suffix);
        let file_temp_path = PathBuf::from(&file_dir).join(&temp_name);
        let mut buf = fs::File::create(&file_temp_path).await?;
        buf.write_all(&data).await?;
        buf.flush().await?;
        drop(buf);

        let id = file_md5(&file_temp_path).await?;

        let mut conn = state.get_db_connection().await?;
        if let Some(existing) = dao::get_file(&mut conn, &id).await? {
            warn!("file {} already exists, returning existing url", id);
            let _ = fs::remove_file(&file_temp_path).await;
            return Ok(existing.url);
        }

        let static_file_dir = format!("{}/{}", STATIC_FILE_DIR, now_date_str);
        fs::create_dir_all(&static_file_dir).await?;

        let static_file_name = format!("{}{}", id, suffix);
        let static_file_path = PathBuf::from(&static_file_dir).join(&static_file_name);
        info!(
            "moving upload file {} to {}",
            file_temp_path.display(),
            static_file_path.display()
        );
        fs::rename(&file_temp_path, &static_file_path).await?;

        let url = format!("/{}/{}", now_date_str, static_file_name);
        let entity = File {
            id: id.clone(),
            filename,
            size,
            url: url.clone(),
            hosting_file: String::new(),
            ctime: Local::now().naive_local(),
        };
        dao::insert_file(&mut conn, &entity).await?;
        Ok(url)
    }

    pub async fn upload_hosting(state: &ApiState, param: HostingUploadParam) -> AppResult<String> {
        let mut conn = state.get_db_connection().await?;

        // 1. check fingerprint: find file by md5 id
        if let Some(existing) = dao::get_file(&mut conn, &param.fingerprint).await? {
            info!(
                "hosting upload: fingerprint {} already exists, returning {}",
                param.fingerprint, existing.url
            );
            return Ok(existing.url);
        }

        // 2. check kind+file_id uniqueness
        if dao::get_hosting_file(&mut conn, &param.kind, &param.file_id)
            .await?
            .is_some()
        {
            return Err(AppError::biz("file alredy existing, overwrite is forbidden"));
        }

        // 3. insert hosting_file
        let extra = param.extra.as_ref().map(|v| v.to_string());
        let hosting = HostingFile {
            kind: param.kind.clone(),
            file_id: param.file_id.clone(),
            extra,
            ctime: Local::now().naive_local(),
        };
        dao::insert_hosting_file(&mut conn, &hosting).await?;

        // 4. insert file
        let file = File {
            id: param.fingerprint.clone(),
            filename: param.filename,
            size: param.size,
            url: param.url.clone(),
            hosting_file: format!("{}:{}", param.kind, param.file_id),
            ctime: Local::now().naive_local(),
        };
        dao::insert_file(&mut conn, &file).await?;

        Ok(param.url)
    }
}
